src/remap/function/compact.rs

```diff
@@ -0,0 +1,406 @@
+            null,
+            string,
+            map,
+            array,
+        }))
+    }
+}
+
+#[derive(Debug, Clone)]
+struct CompactFn {
+    value: Box<dyn Expression>,
+    recursive: Option<Box<dyn Expression>>,
+    null: Option<Box<dyn Expression>>,
+    string: Option<Box<dyn Expression>>,
+    map: Option<Box<dyn Expression>>,
+    array: Option<Box<dyn Expression>>,
+}
+
+impl CompactFn {
+    #[cfg(test)]
+    fn new(
+        value: Box<dyn Expression>,
+        recursive: Option<Box<dyn Expression>>,
+        null: Option<Box<dyn Expression>>,
+        string: Option<Box<dyn Expression>>,
+        map: Option<Box<dyn Expression>>,
+        array: Option<Box<dyn Expression>>,
+    ) -> Self {
+        Self {
+            value,
+            recursive,
+            null,
+            string,
+            map,
+            array,
+        }
+    }
+}
+
+#[derive(Default)]
+struct CompactOptions {
+    recursive: bool,
+    null: bool,
+    string: bool,
+    map: bool,
+    array: bool,
+}
+
+impl CompactOptions {
+    /// Check if the value is empty according to the given options
+    fn is_empty(&self, value: &Value) -> bool {
+        match value {
+            Value::Bytes(bytes) => self.string && bytes.len() == 0,
+            Value::Null => self.null,
+            Value::Map(map) => self.map && map.is_empty(),
+            Value::Array(array) => self.array && array.is_empty(),
+            _ => false,
+        }
+    }
+}
+
+impl Expression for CompactFn {
+    fn execute(&self, state: &mut state::Program, object: &mut dyn Object) -> Result<Value> {
+        let options = CompactOptions {
+            recursive: match &self.recursive {
+                Some(expr) => expr.execute(state, object)?.try_boolean()?,
+                None => false,
+            },
+
+            null: match &self.null {
+                Some(expr) => expr.execute(state, object)?.try_boolean()?,
+                None => false,
+            },
+
+            string: match &self.string {
+                Some(expr) => expr.execute(state, object)?.try_boolean()?,
+                None => false,
+            },
+
+            map: match &self.map {
+                Some(expr) => expr.execute(state, object)?.try_boolean()?,
+                None => false,
+            },
+
+            array: match &self.array {
+                Some(expr) => expr.execute(state, object)?.try_boolean()?,
+                None => false,
+            },
+        };
+
+        match self.value.execute(state, object)? {
+            Value::Map(map) => Ok(Value::from(compact_map(map, &options))),
+            Value::Array(arr) => Ok(Value::from(compact_array(arr, &options))),
+            _ => unreachable!(),
+        }
+    }
+
+    fn type_def(&self, state: &state::Compiler) -> TypeDef {
+        self.value
+            .type_def(state)
+            .fallible_unless(value::Kind::Map | value::Kind::Array)
+            .with_constraint(value::Kind::Map | value::Kind::Array)
```

Comment:
I suppose this line can be removed, such that we rely on whatever the return value of `value` is (which can be either `Map | Array`, or one of the two specifically).